geoips/interfaces/module_based/readers.py

```diff
@@ -19,5 +24,233 @@ class ReadersInterface(BaseModuleInterface):
         "standard": ["metadata_only", "chans", "area_def", "self_register"]
     }
 
+    def read_data_to_xarray_dict(
+        self,
+        fnames,
+        call_single_file_func,
+        metadata_only=False,
+        chans=None,
+        area_def=None,
+        self_register=False,
+    ):
+        """Read in data potentially from multiple scan times into an xarray dict.
+
+        This function does not require that you provide multiple scan times, but allows
+        for that in the case those are provided.
+
+        Call this with information specific to your reader to generate a dictionary of
+        xarray datasets created from the data provided in 'fnames'.
+
+        Parameters
+        ----------
+        fnames : list
+            * List of strings, full paths to files
+        all_metadata : dict
+            * Dictionary of metadata from all files in 'fnames'
+        call_single_file_func : python function
+            * Function which can be used to read a single file from a reader plugin.
+            * Most likely named 'call_single_time'.
+        metadata_only : bool, default=False
+            * Return before actually reading data if True
+        chans : list of str, default=None
+            * List of desired channels (skip unneeded variables as needed).
+            * Include all channels if None.
+        area_def : pyresample.AreaDefinition, default=None
+            * Specify region to read
+            * Read all data if None.
+        self_register : str or bool, default=False
+            * register all data to the specified dataset id (as specified in the
+              return dictionary keys).
+            * Read multiple resolutions of data if False.
+
+        Returns
+        -------
+        dict of xarray.Datasets
+            * dictionary of xarray.Dataset objects with required Variables and
+              Attributes.
+            * Dictionary keys can be any descriptive dataset ids.
+
+        See Also
+        --------
+        :ref:`xarray_standards`
+            Additional information regarding required attributes and variables
+            for GeoIPS-formatted xarray Datasets.
+        """
+        all_metadata = self.concatenate_metadata(
+            [call_single_file_func([x], metadata_only=True)["METADATA"] for x in fnames]
+        )
```

Comment:
Is the ```call_single_file_func``` the ```call``` part of current readers? How would this work with readers that access multiple files at once? Like VIIRS SDR data, where a mix of geo and channel data is read in on one line.

Reply:
``call_single_file_func`` is part of these readers which have been modified in this PR. They all have functions named ``call_single_time``, which is provided as the argument for ``call_single_file_func``. When I took this PR over, it seems like the assumption had been made that one file is enough to get metadata and data for your product, though in retrospect that's not a good assumption to make. We probably could add an argument to ``ReadersInterface.read_data_to_xarray_dict`` which could denote whether or not we want to loop over all files, or loop over lists of files filtered by their timestep. That's probably what I'd recommend if we wanted to change this.
![Screenshot 2024-11-05 at 10 09 20 AM](https://github.com/user-attachments/assets/7f010174-ae94-4b85-b263-760532cdd0b4)

Reply:
If we made that change that would additionally solve this [problem](https://github.com/NRLMMD-GEOIPS/geoips/pull/427#discussion_r1828489853) Mindy noticed.

Reply:
Actually, [here](https://github.com/NRLMMD-GEOIPS/geoips/pull/427#discussion_r1681445713) is where that assumption was made.

Reply:
Could you please create an issue for investigating this, link to this conversation from the issue, link to the issue from this conversation, then you can resolve this conversation?

Reply:
See this [comment](https://github.com/NRLMMD-GEOIPS/geoips/pull/427#discussion_r1831367446) first @mindyls. Depending on what you think, I can either create an issue or update the code match what was said in that comment.

Reply:
See this [comment](https://github.com/NRLMMD-GEOIPS/geoips/pull/427#discussion_r1831802202). This should be fixed now.